src/fucking_shit.cu

```diff
@@ -153,38 +156,68 @@ __host__ __device__ MapNode *find_nearest_mapnode(const Polyhedron *const polyhe
     return find_nearest_mapnode_greedy(dest, dest_face->get_node());
 }
 
+#ifdef COMPILE_FOR_CPU
+
+base_atomic_type atomicCAS(base_atomic_type *address, const base_atomic_type compare, const base_atomic_type val)
+{
+    base_atomic_type ans = *address;
+
+    if(*address == compare)
+        *address = val;
+
+    return ans;
+}
+
+base_atomic_type atomicAdd(base_atomic_type *address, const base_atomic_type value)
+{
+    base_atomic_type ans = *address;
+    *address += value;
+    return ans;
+}
+
+double atomicAdd(double *address, const double value)
+{
+    double ans = *address;
+    *address += value;
+    return ans;
+}
 
-// The following code is from https://stackoverflow.com/a/62094892/11248508 (@kolayne can explain)
-// `address` CANNOT be pointer to const, because we are trying to edit memory by it's address
+#endif //COMPILE_FOR_CPU
+
+// The following code is compied from https://stackoverflow.com/a/62094892/11248508 and modified (@kolayne can explain)
+// `address` CANNOT be pointer to const, because we are trying to edit memory by it's source
 __device__ bool atomicCAS(bool *const address, const bool compare, const bool val)
 {
     static_assert(sizeof(base_atomic_type) > 1, "The local atomicCAS implementation won't work if `base_atomic_type"
                                                    "type size <= 1");
     static_assert((sizeof(base_atomic_type) - 1 & sizeof(base_atomic_type)) == 0,
             "The local atomicCAS implementation won't work if `base_atomic_type` type size is not a power of 2");
 
-    
-    auto addr = (base_atomic_type)address;
-    base_atomic_type pos = addr & (sizeof(base_atomic_type) - 1);  // byte position within the int
-    auto *int_addr = (base_atomic_type *)(addr - pos);  // int-aligned address
-    base_atomic_type old = *int_addr, assumed, ival;
+
+    auto address_num = (unsigned long long)address;
```

Comment:
Why `unsigned long long` instead of `base_atomic_type`?

Reply:
Hi! The types of this variable (`address_num`) and the variable from the next line (`pos`) were changed because their types shouldn't be anyhow related to `base_atomic_type`. `address_num` is a variable representing an address of another variable as a number. So it doesn't matter which type does the other variable have, the address should always be `unsigned long long`. Almost the same explanation with `pos` - it should always be a number either